### DISCUSSION TITLE
Collaboration Notes

### DISCUSSION THREAD

**Discussion by NTDuck** [NTDuck/tomfoolery#1] [Announcements]
### Overview
1. Uses [Google Java Style Guide](https://google.github.io/styleguide/javaguide.html).
2. Uses [Gradle](https://gradle.org/) over [Maven](https://maven.apache.org/index.html) (for [increased productivity](https://tomgregory.com/gradle/maven-vs-gradle-comparison/))
3. Implements [Clean Architecture](https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html). Might also consider adopting [DDD](https://github.com/gg-daddy/ebooks/blob/c063174df5368b195671834c020f707e1783f0d9/Eric%20Evans%202003%20-%20Domain-Driven%20Design%20-%20Tackling%20Complexity%20in%20the%20Heart%20of%20Software.pdf)
4. *Please* keep the code [clean](https://github.com/martinmurciego/good-books/blob/0a331e0beb2b93328622fdaf546be04e97419151/Clean%20Code_%20A%20Handbook%20of%20Agile%20Software%20Craftsmanship%20-%20Robert%20C.%20Martin.pdf)

#### Preface
[The Dependency Rule](https://khalilstemmler.com/wiki/dependency-rule/) (source code dependencies can only point inwards) should always apply.

#### `core/domain` (Entities)
> [...] Entities are the business objects of the application. They encapsulate the most general and high-level rules. They are the least likely to change when something external changes. [...] No operational change to any particular application should affect the entity layer.

We use entities as if they're POJOs. As of now, the application has two major entities of concern: `Dictionary` and `DictionaryEntry`. A user (for now, an instance of the application) can contain many `Dictionary`. A `Dictionary` consists of many (as many as 470,000) `DictionaryEntry`. A `DictionaryEntry` typically contain necessary data such as headword or definition. Two `DictionaryEntry` is equal if and only if their headword is identical.

Therefore, according to DDD terminologies, `Dictionary` would fall into the `AggregateRoot` category, and `DictionaryEntry` would fall into the `Entity` category.

#### `core/usecases`
> The software in this layer contains application specific business rules. It encapsulates and implements all of the use cases of the system. These use cases orchestrate the flow of data to and from the entities, and direct those entities to use their enterprise wide business rules to achieve the goals of the use case. [...] We do not expect changes in this layer to affect the entities. We also do not expect this layer to be affected by changes to externalities such as the database, the UI, or any of the common frameworks. This layer is isolated from such concerns. [...] We do, however, expect that changes to the operation of the application will affect the use-cases and therefore the software in this layer. If the details of a use-case change, then some code in this layer will certainly be affected.

About the business/application-specific rules:
1. The application has an immutable `Dictionary` (which comes by default) and a collection of mutable `Dictionary` (which is empty by default)
2. Regarding the collection of mutable `Dictionary`, the application supports the following use cases:
- `AddDictionaryUseCase`: Import a `Dictionary` from a file and add it to the collection.
- `UpdateDictionaryUseCase`: Import a `Dictionary` from a file and replace the current selected `Dictionary` in the collection.
- `DeleteDictionaryUseCase`: Remove the selected `Dictionary` in the collection.
- `SelectDictionaryUseCase`: Make an existing `Dictionary` in the collection becomes selected.
- `ShowDictionariesUseCase`: Display all `Dictionary` in the collection, including the selected.
3. Regarding any mutable `Dictionary`, the application supports the following use cases:
- `AddDictionaryEntryUseCase`: Add a `DictionaryEntry` to the `Dictionary` based on user input. Most probably a html-like form.
- `UpdateDictionaryEntryUseCase`: Replace a `DictionaryEntry` in the `Dictionary` with a new `DictionaryEntry` with identical headword, based on user input. Throws a `NotFoundException` if no `DictionaryEntry` with identical headword is found in the `Dictionary`.
- `DeleteDictionaryEntryUseCase`: Remove a `DictionaryEntry` from the `Dictionary`. Throws a `NotFoundException` if no `DictionaryEntry` with identical headword is found in the `Dictionary`.
4. Regarding any `Dictionary` (regardless of mutability), the application supports the following use cases:
- `GetDictionaryEntryUseCase`: Fetch all `DictionaryEntry` from all `Dictionary` with matching headword (exact match).
- `SearchDictionaryEntriesUseCase`: Fetch all `DictionaryEntry` from all `Dictionary` with headwords that matches the inputted prefix.
- `ShowDictionaryEntriesUseCase`: Fetch all `DictionaryEntry` from all `Dictionary`. Might have limit imposed + pagination (?).
5. Game is intended to be hangman.
 
#### `core/dataproviders`
Exists to ensure strict adherence to The Dependency Rule. Define interfaces that work with entities. Very commonly, (almost) every use case could be observed having a reference to a dataprovider.

No implementation of dataproviders should be provided. They only need to be sufficiently operational for use cases.

#### `infrastructures/dataproviders` (Interface Adapters)
> The dataproviders layer primarily focuses on mapping database access. However, it can also encompass [API](https://developers.redhat.com/topics/api-management/) access. [...] The rationale behind this is that a dataprovider's purpose is to retrieve data, regardless of whether it originates from a database or an external service.

For example, if a `DictionaryEntryRepository` is defined in `core/dataproviders`, then this layer might implement `InMemoryDictionaryEntryRepsitory`, `MySQLDictionaryEntryRepository`, `WordsAPIDictionaryEntryRepository`, etc.

#### `configurations` (Frameworks and Drivers)
> Here, in this section, we have the "glue" that connects and wires the different layers of our application. It involves making decisions about the infrastructure to use, selecting use cases, determining which endpoints to expose, and choosing the appropriate technologies.

We have decided to put all the Views, Presenters, and Controllers in this layer, instead of the infrastructures layer (as Clean Architecture might have suggested) for we believe that such components might vary significantly between e.g. UI frameworks. Also, for now, we put the entry point of the application i.e. the one class that has the `main()` method because we feel like doing so.

#### Coda
Every layer might have e.g. the `utils` package inside it somewhere to provide additional utilities.

**Comment by NTDuck** [comment]
**Old requirements.txt**:
```

// Comes when installation, user can't modify
Repo default;

// = 0 when installation, user can CRUD
List<Repo> userRepos;

// operate on the current repo
Add, Update, Delete... DictionaryEntryUseCases

// operate on all repos
Get, SearchByPrefix, Search

// operate on repos
CreateRepo, UpdateRepo, DeleteRepo, SelectRepo
ImportFromFileUseCase: add new repository to userRepos i.e. CreateRepo
ExportFromFileUseCase: export current repository to local file

// Game: todo

// impls
Db/API/... Repo: EtoE
EtoV: translate all E definitions to V
```

**Comment by adnope** [comment]
sick!

**Reply by NTDuck** [reply]
spyep

**Comment by NTDuck** [comment]
### Update
The application will lean towards more DDD and less Clean Architecture; specifically, use cases will be replaced with domain methods.

**Reply by adnope** [reply]
real

**Reply by NTDuck** [reply]
### Update
Only 1 dictionary, to simplify things.
